### PR TITLE
Fixes Compile warnings and Adds Panic and Compile time and runtime controlled Debugging

### PR DESCRIPTION
**Pull request by davids5**
This adds Debug logging and Panic features.

``` cpp
/*
 * This module supports runtime and compile time message filtering.
 *
 * For a message be compiled in the message type has to be >= to then the LOG_LEVEL_AT_COMPILE_TIME
 * For a message to be output the message type has to be >= then the LOG_LEVEL_AT_RUN_TIME
 *
 * This module assumes one of two #defines RELEASE_BUILD or DEBUG_BUILD exist
 * Then the LOG_LEVEL_AT_COMPILE_TIME and LOG_LEVEL_AT_RUN_TIME are set based on RELEASE or DEBUG
 * For a release build
 *     LOG_LEVEL_AT_COMPILE_TIME = WARN_LEVEL
 *         LOG_LEVEL_AT_RUN_TIME = WARN_LEVEL
 * For a debug build:
 *     LOG_LEVEL_AT_COMPILE_TIME = LOG_LEVEL
 *     LOG_LEVEL_AT_RUN_TIME = LOG_LEVEL
 * The API
 * LOG(fmt,...)
 * DEBUG(fmt,...)
 * WARN(fmt,...)
 * ERROR(fmt,...)
 * PANIC(code,fmt,...)
 */
```

Usage:

``` cpp
define DEBUG_BUILD=y on make command line and add the following to your Application.cpp

void debug_output_(const char *p)
{
  static boolean once = false;
 if (!once)
   {
     once = true;
     Serial1.begin(115200);
   }

 Serial1.print(p);
}
The API looks like

LOG("Want %d more cores",count);
WARN("Running %s on cores only %d more left",,"Low",count);
DEBUG("connection closed %d",retValue);
ERROR("Flash write Failed @0x%0x",badd_address);
PANIC(HardFault,"Hit Hardfault");
```

Panic codes are defined in panic_codes.h

PANIC(code,[text])

``` cpp
// Add a panic code def_panic_codes(_class, led, code)
// the panic module will send SOS followed by 900 ms delay
// followed by 300 blinks of the value of code
/// ...---...    code   ...---...

This is added to the while(1) loop in the code to help identify a Fault or Fault condition
Currently used

def_panic_codes(Faults,RGB_COLOR_RED,HardFault)      1 blink
def_panic_codes(Faults,RGB_COLOR_RED,NMIFault)        2 etc....
def_panic_codes(Faults,RGB_COLOR_RED,MemManage)
def_panic_codes(Faults,RGB_COLOR_RED,BusFault)
def_panic_codes(Faults,RGB_COLOR_RED,UsageFault)

def_panic_codes(Cloud,RGB_COLOR_RED,InvalidLenth)

def_panic_codes(System,RGB_COLOR_RED,Exit)
def_panic_codes(System,RGB_COLOR_RED,OutOfHeap)


```

### CODE REVIEW

File: SPARK_Firmware_Driver/src/cc3000_spi.c [hunk 1]
```diff
@@ -482,97 +343,98 @@ void SpiContReadOperation(void)
...
+                         {
+                           /* Odd so make it even */
+
+                           data_to_recv++;
+                         }
+
+                       /* Read the whole payload in at the beginning of the buffer
+                        * Will it fit?
+                        */
+                       SPARK_ASSERT(data_to_recv >= arraySize(wlan_rx_buffer));
```
Comment (towynlin) [inline SPARK_Firmware_Driver/src/cc3000_spi.c]: @davids5 Shouldn't this assertion be the other way around?

**Comment by davids5** [comment]
Yes, there was 2 bugs! Fixed in e0fea59

**Comment by towynlin** [comment]
OK, thanks, looks good. However, now that I'm trying to test, I get a linker error because you call `millis()` in a few places, which is defined in core-firmware.

```
Building target: core-firmware.elf
Invoking: ARM GCC C++ Linker
arm-none-eabi-g++ -g3 -gdwarf-2 -Os -mcpu=cortex-m3 -mthumb  -I../inc -I../../core-common-lib/CMSIS/Include -I../../core-common-lib/CMSIS/Device/ST/STM32F10x/Include -I../../core-common-lib/STM32F10x_StdPeriph_Driver/inc -I../../core-common-lib/STM32_USB-FS-Device_Driver/inc -I../../core-common-lib/CC3000_Host_Driver -I../../core-common-lib/SPARK_Firmware_Driver/inc -I../../core-communication-lib/lib/tropicssl/include -I../../core-communication-lib/src -I. -ffunction-sections -Wall -fmessage-length=0 -MD -MP -MF core-firmware.elf.d -DUSE_STDPERIPH_DRIVER -DSTM32F10X_MD -DDFU_BUILD_ENABLE  obj/src/application.o obj/src/main.o obj/src/newlib_stubs.o obj/src/spark_utilities.o obj/src/spark_wiring.o obj/src/spark_wiring_i2c.o obj/src/spark_wiring_interrupts.o obj/src/spark_wiring_ipaddress.o obj/src/spark_wiring_network.o obj/src/spark_wiring_print.o obj/src/spark_wiring_servo.o obj/src/spark_wiring_spi.o obj/src/spark_wiring_stream.o obj/src/spark_wiring_string.o obj/src/spark_wiring_tcpclient.o obj/src/spark_wiring_tcpserver.o obj/src/spark_wiring_udp.o obj/src/spark_wiring_usartserial.o obj/src/spark_wiring_usbserial.o obj/src/spark_wlan.o obj/src/stm32_it.o obj/src/usb_desc.o obj/src/usb_endp.o obj/src/usb_istr.o obj/src/usb_prop.o obj/src/wifi_credentials_reader.o obj/startup/startup_stm32f10x_md.o --output core-firmware.elf -T../linker/linker_stm32f10x_md_dfu.ld -nostartfiles -Xlinker --gc-sections  -L../../core-common-lib/build -lcore-common-lib -L../../core-communication-lib/build -lcore-communication-lib -Wl,-Map,core-firmware.map
../../core-common-lib/build/libcore-common-lib.a(evnt_handler.o): In function `hci_event_handler':
/Users/zac/code/spark/core-common-lib/build/../CC3000_Host_Driver/evnt_handler.c:229: undefined reference to `millis'
/Users/zac/code/spark/core-common-lib/build/../CC3000_Host_Driver/evnt_handler.c:234: undefined reference to `millis'
../../core-common-lib/build/libcore-common-lib.a(debug.o): In function `log_print_':
/Users/zac/code/spark/core-common-lib/build/../SPARK_Firmware_Driver/src/debug.c:34: undefined reference to `millis'
collect2: error: ld returned 1 exit status
make: *** [core-firmware.elf] Error 1
```

Grepping for millis in the common library gets those 3 results, plus the extern declaration:

```
./CC3000_Host_Driver/evnt_handler.c:229:    volatile int32_t start = millis();
./CC3000_Host_Driver/evnt_handler.c:234:                    volatile int32_t now = millis();
./SPARK_Firmware_Driver/src/debug.c:16:extern unsigned long millis(void);
./SPARK_Firmware_Driver/src/debug.c:34:        int trunc = snprintf(_buffer, arraySize(_buffer), "%010lu:<%s> %s %s(%d):", millis(), levels[level], func, file, line);
```

The common lib can't depend on the core-firmware repo, since the common lib is used by, e.g., the bootloader and the cc3000-patch-programmer.  Please submit some solution here.

**Comment by davids5** [comment]
Is there a time source in lib common?

**Comment by davids5** [comment]
Are you available on Skype today?

**Comment by davids5** [comment]
Would this work: 

Add the following to to the core-common-lib:

``` cpp
void System1MsTick(void) . 
{
   system_1ms_tick++;
}
uint32_t GetSystem1MsTick()
{
 return system_1ms_tick;
}
```

static uint32_t system_1ms_tick = 0.

 System1MsTick() will be defined weak and called off of the SysTick_Handler from the lib users

millis(); will use GetSystem1MsTick

And  the code in common lib will use  GetSystem1MsTick

Thoughts?

**Comment by towynlin** [comment]
OK @davids5 — Now the 1ms tick PRs are merged, so update this one (and the firmware one) to use that code. Thanks!

**Comment by davids5** [comment]
OK @towynlin merged compiles and runs!

**Pull request merged**